File: Ground/Ground/QuoteService.swift
import Foundation
import Combine
import SwiftUI

struct Quote: Codable, Identifiable {
    var id: String { q }
    let q: String
    let a: String
}

@MainActor
class QuoteService: ObservableObject {
    static let shared = QuoteService()

    private let zenQuotesURL = URL(string: "https://zenquotes.io/api/random")!
    private let cacheKey = "ground_quote_cache"

    @Published var currentQuote: Quote?

    private init() {}

    private func localDateString() -> String {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return f.string(from: Date())
    }

    func getQuoteOfTheDay() async -> Quote {
        let today = localDateString()

        if let cachedData = UserDefaults.standard.data(forKey: cacheKey),
           let cached = try? JSONDecoder().decode(QuoteCache.self, from: cachedData),
           cached.date == today {
            currentQuote = cached.quote
            return cached.quote
        }

        do {
            let (data, _) = try await URLSession.shared.data(from: zenQuotesURL)
            if let quotes = try? JSONDecoder().decode([Quote].self, from: data), let quote = quotes.first {
                saveToCache(quote: quote, date: today)
                currentQuote = quote
                return quote
            }
        } catch {
            print("Failed to fetch quote: \(error)")
        }

        let fallback = Quote(q: "Stay grounded.", a: "Ground")
        currentQuote = fallback
        return fallback
    }


    func shouldShowToday() -> Bool {
        UserDefaults.standard.string(forKey: "ground_quote_last_shown") != localDateString()
    }

    func markShownToday() {
        UserDefaults.standard.set(localDateString(), forKey: "ground_quote_last_shown")
    }

    private func saveToCache(quote: Quote, date: String) {
        if let encoded = try? JSONEncoder().encode(QuoteCache(date: date, quote: quote)) {
            UserDefaults.standard.set(encoded, forKey: cacheKey)
        }
    }

    struct QuoteCache: Codable {
        let date: String
        let quote: Quote
    }
}

struct Profile: Codable {
    let id: UUID
    let quote_start_time: String?
    let last_quote_shown_at: String?
    let username: String?
}

File: ground-app/src/services/QuoteService.js
const FALLBACK_QUOTE = { q: "Stay grounded.", a: "Ground" }
const ZEN_QUOTES_API = 'https://zenquotes.io/api/today'

export const QuoteService = {
  async getQuoteOfTheDay() {
    const today = new Date().toISOString().split('T')[0]
    const cached = localStorage.getItem('ground_quote_cache')

    if (cached) {
      const { date, quote } = JSON.parse(cached)
      if (date === today) return quote
    }

    try {
      const response = await fetch(ZEN_QUOTES_API)
      const data = await response.json()
      if (data && data[0]) {
        const quote = { q: data[0].q, a: data[0].a }
        localStorage.setItem('ground_quote_cache', JSON.stringify({ date: today, quote }))
        return quote
      }
    } catch {
      // API down — use fallback
    }

    return FALLBACK_QUOTE
  },

  shouldShowModal(profile) {
    if (!profile) return false

    const { quote_start_time, last_quote_shown_at } = profile
    const now = new Date()
    const [hours, minutes] = (quote_start_time || '06:00:00').split(':').map(Number)

    let referencePoint = new Date()
    referencePoint.setHours(hours, minutes, 0, 0)
    if (now < referencePoint) referencePoint.setDate(referencePoint.getDate() - 1)

    if (!last_quote_shown_at) return true
    return new Date(last_quote_shown_at) < referencePoint
  },

  async markQuoteAsShown(supabase, userId) {
    await supabase
      .from('profiles')
      .update({ last_quote_shown_at: new Date().toISOString() })
      .eq('id', userId)
  },
}
